### PR TITLE
🎁 Add `search` to file location methods in tasks

### LINKED ISSUE #160771: The default `"fileLocation": "relative"` property for tasks does not work with npm tasks that use the `"path"` property
Does this issue occur when all extensions are disabled?: No

- VS Code Version: 1.71.0
- OS Version: macOS 12.15.1

Steps to Reproduce:

1. Define any npm task that is in a subfolder using the `"path"` property, and define a `problemMatcher` that uses `"fileLocation": "relative"`.
2. It won't work, as the directory that it should be relative to will be the `${workspaceFolder}` rather than the actual working directory of the npm script set by `"path"`.

### PR DESCRIPTION
**Pull request by babakks**
Fixes #160771

A new `fileLocation` method, named `search`, is added to problem matches. The usage of this file location method is as:

```jsonc
"problemMatcher": {
    "fileLocation": [
        "search",
        {
            "include": [ // (Required)
                "${workspaceFolder}/subdir1",
                "${workspaceFolder}/subdir2"
            ],
            "exclude": [ // (Optional)
                "${workspaceFolder}/subdir1/subdir-to-ignore",
                "${workspaceFolder}/subdir2/subdir-to-ignore"
            ]
        }
    ]
```

The `include` field is required, but the `exclude` is not.

To verify this PR, you can use [babakks/vscode-verify-search-file-location](https://github.com/babakks/vscode-verify-search-file-location) repo as the test workspace.

### CODE REVIEW

**Review by meganrogge — DISMISSED** [review]
Very nice 👏🏼

**Review by alexr00 — COMMENTED** [review]
I didn't see variables in the `"includes"` and `"excludes"` section get resolved when testing this.

File: src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts [hunk 1]
```diff
@@ -1719,7 +1725,15 @@ export class TerminalTaskSystem extends Disposable implements ITaskSystem {
 					copy.uriProvider = taskSystemInfo.uriProvider;
 				}
 				if (hasFilePrefix) {
-					copy.filePrefix = await this._resolveVariable(resolver, copy.filePrefix);
+					if (Types.isString(matcher.filePrefix)) {
+						copy.filePrefix = await this._resolveVariable(resolver, copy.filePrefix as string);
```
[the remark this answers is not in the document]
Comment (babakks) [reply src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts]: Applied.

File: src/vs/workbench/contrib/tasks/common/problemMatcher.ts [hunk 2]
```diff
@@ -212,6 +215,18 @@ export async function getResource(filename: string, matcher: ProblemMatcher, fil
 
 		matcherClone.fileLocation = FileLocationKind.Absolute;
 		return getResource(filename, matcherClone);
+	} else if (kind === FileLocationKind.Search && fileService) {
+		const fsProvider = fileService.getProvider('file');
```
[the remark this answers is not in the document]
Comment (babakks) [reply src/vs/workbench/contrib/tasks/common/problemMatcher.ts:220]: Applied.

File: src/vs/workbench/contrib/tasks/common/problemMatcher.ts [hunk 3]
```diff
@@ -1612,9 +1678,24 @@ export namespace Schemas {
...
+									'exclude': { type: 'array', items: { type: 'string' } },
+								},
+								required: ['include']
+							}
+						],
+						additionalItems: false,
 					}
 				],
-				description: localize('ProblemMatcherSchema.fileLocation', 'Defines how file names reported in a problem pattern should be interpreted. A relative fileLocation may be an array, where the second element of the array is the path the relative file location.')
+				description: localize('ProblemMatcherSchema.fileLocation', 'Defines how file names reported in a problem pattern should be interpreted. A relative fileLocation may be an array, where the second element of the array is the path of the relative file location.')
```
[the remark this answers is not in the document]
Comment (babakks) [reply src/vs/workbench/contrib/tasks/common/problemMatcher.ts]: Added to the description.

File: src/vs/workbench/contrib/tasks/common/problemMatcher.ts [hunk 4]
```diff
@@ -1612,9 +1678,24 @@ export namespace Schemas {
...
 						}
+					},
+					{
+						type: 'array',
+						items: [
+							{ type: 'string', enum: ['search'] },
+							{
+								type: 'object',
+								properties: {
+									'include': { type: 'array', items: { type: 'string' } },
```
[the remark this answers is not in the document]
Comment (babakks) [reply src/vs/workbench/contrib/tasks/common/problemMatcher.ts]: Added examples to the schema which will be rendered as suggestions in the runtime.

Also, for a better experience, changed the type of `include`/`exclude` arrays to accept single values as well; i.e., they're now `string | string[]`.

**Comment by babakks** [comment]
@alexr00 I've also added two other examples for the `relative` and `autoDetect` file location methods.

**Comment by babakks** [comment]
> I didn't see variables in the `"includes"` and `"excludes"` section get resolved when testing this.

@alexr00 I'm not sure I'm getting your point, correctly. When I tested, the variables (e.g., `"${workspaceFolder}"`) were replaced with their values via the `TerminalTaskSystem._resolveVariable` method.

**Comment by babakks** [comment]
@meganrogge @alexr00 Couldn't ask you both for the review from the side links, so could you please re-review this?

**Comment by babakks** [comment]
@meganrogge @alexr00 We can also make the `include` property optional (default to `${workspaceFolder}`). What do you think?

File: src/vs/workbench/contrib/tasks/common/problemMatcher.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (babakks) [reply src/vs/workbench/contrib/tasks/common/problemMatcher.ts]: But, I'm now sure why validation didn't work.

**Comment by babakks** [comment]
It had a shortcoming that when the `filename` was more than a mere filename and extension (e.g., 'subdir/name.ext`). Fixed that in the last commit; also updated the verification repo to include new tasks for this purpose.

**Comment by meganrogge** [comment]
> @meganrogge @alexr00 We can also make the `include` property optional too (default to `${workspaceFolder}`). What do you think?

I like that idea - thanks for working on this

**Comment by babakks** [comment]
The `include` property is now optional and its default is `"${workspaceFolder}"`.

**Review by meganrogge — APPROVED** [review]

**Pull request merged**